File: GLLara/GLLPoseExporter.m
//
//  GLLPoseExporter.m
//  GLLara
//
//  Created by Torsten Kammer on 31.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLPoseExporter.h"

#import "GLLItem.h"
#import "GLLItemBone.h"
#import "GLLModelBone.h"

@interface GLLPoseExporter ()

@property (nonatomic) id bonesList;

@end

@implementation GLLPoseExporter

- (id)initWithItem:(GLLItem *)item
{
	return [self initWithBones:item.bones];
}

- (id)initWithBones:(id)bones
{
	NSParameterAssert(bones);
	if (!(self = [super init])) return nil;
	
	self.bonesList = bones;
	
	return self;
}

- (NSString *)poseDescription
{
	NSMutableString *string = [[NSMutableString alloc] init];
	
	for (GLLItemBone *bone in self.bonesList)
	{
		if (self.skipUnused && [bone.bone.name hasPrefix:@"unused"])
			continue;
		
		[string appendFormat:@"%@: %f %f %f %f %f %f\r\n", bone.bone.name, bone.rotationX * 180.0 / M_PI, bone.rotationY * 180.0 / M_PI, bone.rotationZ * 180.0 / M_PI, bone.positionX, bone.positionY, bone.positionZ];
	}
	
	return [string copy];
}

@end

File: GLLara/GLLPoseExporter.swift
//
//  GLLPoseExporter.m
//  GLLara
//
//  Created by Torsten Kammer on 31.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

import Foundation

/*!
 * @abstract Exports a pose for a complete item or just a few bones.
 * @discussion Since it can export a pose for only a few bones, this
 * functionality does not belong to the item.
 */
@objc class GLLPoseExporter: NSObject {
    @objc convenience init(item: GLLItem) {
        self.init(bones: item.combinedBones()!.map { $0 as! GLLItemBone })
    }
    
    @objc init(bones: [GLLItemBone]) {
        bonesList = bones
        
        super.init()
    }
    
    let bonesList: [GLLItemBone]
    
    @objc var skipUnused = false
    
    @objc var poseDescription: String {
        var result = ""
        for bone in bonesList {
            if skipUnused && bone.bone.name.hasPrefix("unused") {
                continue
            }
            
            result += "\(bone.bone.name): \(bone.rotationX * 180/Float.pi) \(bone.rotationY * 180/Float.pi) \(bone.rotationZ * 180/Float.pi) \(bone.positionX) \(bone.positionY) \(bone.positionZ)\r\n"
        }
        return result
    }
}
